### PR TITLE
Fix windows compile failure with clang/gcc

### PR DESCRIPTION
**Pull request by Sixmorphugus**
clang++.exe (not clang-cl.exe) is my preferred C++ compiler on windows. I noticed this wasn't working for a couple of reasons;

- If you are using clang or gcc, the CMakeLists.txt files for DiligentCore simply assume you are on Linux without checking and link to libdl.

I fixed this by adding a PLATFORM_WIN32 branch before doing that.

- ShaderTools produces a ton of compile errors when you try to compile its Dx parts on Windows with clang - Windows MSVC (as recommended by the build instructions, I did read them) does not have any complaints. I can only assume all the compile errors are here because you never compile this code if you're not on windows, as all the issues here are absent in the rest of the codebase;
    - switch statements that don't cover every case
    - overriding of functions from Windows SDK with different exception spec 
    - unused const globals

I fixed these by adding a PLATFORM_WIN32 branch to disable the windows-only microsoft-exception-spec warning ~~(and also all unused warnings since there was an unused variable in ShaderTools too)~~ then made fixes to the code to fix the other things. ~~Disabling unused warnings may not be acceptable,~~ open to feedback. Notably I'm uneasy about making those default switch statements that do nothing but I don't know what was intended by the existing apparent undefined behavior.

This is all it took to get it working for my project but I'm sure you guys will have thoughts.

### CODE REVIEW

**Comment by Sixmorphugus** [comment]
Let me fix the formatting issue.

**Comment by Sixmorphugus** [comment]
I just tried to compile the D3D12 backend and a LOT more things are broken. Looking into it....

**Comment by Sixmorphugus** [comment]
All compiles under winclang now but again a lot had to be changed so will probably need review. You don't have to accept this, am happy just to maintain my own branch that compiles under clang.

File: Graphics/GraphicsEngineD3D11/include/ShaderVariableManagerD3D11.hpp [hunk 1]
```diff
@@ -138,7 +138,7 @@ class ShaderVariableManagerD3D11 : ShaderVariableManagerBase<EngineD3D11ImplTrai
             ShaderVariableD3D11Base<ConstBuffBindInfo, D3D11_RESOURCE_RANGE_CBV>{ParentLayout, ResIndex}
         {}
 
-        __forceinline void BindResource(const BindResourceInfo& BindInfo);
+        void BindResource(const BindResourceInfo& BindInfo);
```
Comment (TheMostDiligent) [inline Graphics/GraphicsEngineD3D11/include/ShaderVariableManagerD3D11.hpp:141]: Why does `__forceinline` need to be removed here?
It should be defined as inline.
https://github.com/DiligentGraphics/DiligentCore/blob/master/CMakeLists.txt#L424

File: Graphics/GraphicsEngineD3D12/src/RenderDeviceD3D12Impl.cpp [hunk 2]
```diff
@@ -283,11 +283,11 @@ RenderDeviceD3D12Impl::~RenderDeviceD3D12Impl()
     ReleaseStaleResources(true);
 
 #ifdef DILIGENT_DEVELOPMENT
-    for (auto i = 0; i < _countof(m_CPUDescriptorHeaps); ++i)
+    for (Uint64 i = 0; i < _countof(m_CPUDescriptorHeaps); ++i)
```
Comment (TheMostDiligent) [inline Graphics/GraphicsEngineD3D12/src/RenderDeviceD3D12Impl.cpp:286]: size_t is better here

File: Graphics/GraphicsEngineD3D12/src/RenderDeviceD3D12Impl.cpp [hunk 3]
```diff
@@ -283,11 +283,11 @@ RenderDeviceD3D12Impl::~RenderDeviceD3D12Impl()
...
     {
         DEV_CHECK_ERR(m_CPUDescriptorHeaps[i].DvpGetTotalAllocationCount() == 0, "All CPU descriptor heap allocations must be released");
     }
-    for (auto i = 0; i < _countof(m_GPUDescriptorHeaps); ++i)
+    for (Uint64 i = 0; i < _countof(m_GPUDescriptorHeaps); ++i)
```
Comment (TheMostDiligent) [inline Graphics/GraphicsEngineD3D12/src/RenderDeviceD3D12Impl.cpp:290]: size_t

File: Graphics/GraphicsEngineD3DBase/include/EngineFactoryD3DBase.hpp [hunk 4]
```diff
@@ -110,12 +109,10 @@ class EngineFactoryD3DBase : public EngineFactoryBase<BaseInterface>
 
         UINT numModes = 0;
         // Get the number of elements
-        auto hr = pOutput->GetDisplayModeList(DXIGFormat, 0, &numModes, NULL);
```
Comment (TheMostDiligent) [inline Graphics/GraphicsEngineD3DBase/include/EngineFactoryD3DBase.hpp:113]: Why was this removed?

File: Graphics/GraphicsEngineD3DBase/include/EngineFactoryD3DBase.hpp [hunk 5]
```diff
@@ -110,12 +109,10 @@ class EngineFactoryD3DBase : public EngineFactoryBase<BaseInterface>
...
         if (DisplayModes != nullptr)
         {
             // Get the list
             std::vector<DXGI_MODE_DESC> DXIDisplayModes(numModes);
-            hr = pOutput->GetDisplayModeList(DXIGFormat, 0, &numModes, DXIDisplayModes.data());
```
Comment (TheMostDiligent) [inline Graphics/GraphicsEngineD3DBase/include/EngineFactoryD3DBase.hpp:118]: Why was this removed?

File: Graphics/HLSL2GLSLConverterLib/include/GLSLDefinitions.h [hunk 6]
```diff
@@ -34,9 +34,9 @@
 // From GLES 3.1 spec:
```
Comment (TheMostDiligent) [inline Graphics/HLSL2GLSLConverterLib/include/GLSLDefinitions.h:34]: Let's not touch this file - it only adds a lot of changes to the PR without real need.

**Review by TheMostDiligent — CHANGES_REQUESTED** [review]

File: Graphics/GraphicsEngineD3D11/include/ShaderVariableManagerD3D11.hpp (hunk 1 above)
Comment (Sixmorphugus) [reply Graphics/GraphicsEngineD3D11/include/ShaderVariableManagerD3D11.hpp:141]: For some reason it doesn't work in clang. Clang says it's not implemented if I leave it as is. I will double check if that define is going through.

File: Graphics/HLSL2GLSLConverterLib/include/GLSLDefinitions.h (hunk 6 above)
Comment (Sixmorphugus) [reply Graphics/HLSL2GLSLConverterLib/include/GLSLDefinitions.h:34]: Not sure how this happened, will remove.

File: Graphics/GraphicsEngineD3DBase/include/EngineFactoryD3DBase.hpp (hunk 4 above)
Comment (Sixmorphugus) [reply Graphics/GraphicsEngineD3DBase/include/EngineFactoryD3DBase.hpp:113]: It seemed to be unused and was causing an unused variable warning. I thought it was dead leftover code so I removed it. Can put it back.

File: Graphics/GraphicsEngineD3DBase/include/EngineFactoryD3DBase.hpp (hunk 5 above)
Comment (Sixmorphugus) [reply Graphics/GraphicsEngineD3DBase/include/EngineFactoryD3DBase.hpp:118]: Same as above

File: Graphics/GraphicsEngineD3DBase/include/EngineFactoryD3DBase.hpp (hunk 4 above)
Comment (TheMostDiligent) [reply Graphics/GraphicsEngineD3DBase/include/EngineFactoryD3DBase.hpp:113]: The function `GetDisplayModeList` does what the name suggests - it returns the list of display modes through output parameters.
Removing the function will break display mode enumeration.

File: Graphics/GraphicsEngineD3D11/include/ShaderResourceCacheD3D11.hpp [hunk 7]
```diff
@@ -375,7 +375,15 @@ class ShaderResourceCacheD3D11 : public ShaderResourceCacheBase
...
-    void TransitionResourceStates(DeviceContextD3D11Impl& Ctx);
+    void TransitionResourceStates(DeviceContextD3D11Impl& Ctx)
+    {
+        VERIFY_EXPR(IsInitialized());
+        
+        TransitionResources<Mode>(Ctx, static_cast<ID3D11Buffer*>(nullptr));
+        TransitionResources<Mode>(Ctx, static_cast<ID3D11ShaderResourceView*>(nullptr));
+        TransitionResources<Mode>(Ctx, static_cast<ID3D11SamplerState*>(nullptr));
+        TransitionResources<Mode>(Ctx, static_cast<ID3D11UnorderedAccessView*>(nullptr));
+    }
```
[the remark this answers is not in the document]
Comment (Sixmorphugus) [reply Graphics/GraphicsEngineD3D11/include/ShaderResourceCacheD3D11.hpp:386]: The error was:
```In file included from D:/Repos/funeral/extlibs/DiligentCore/Graphics/GraphicsEngineD3D11/src/ShaderResourceBindingD3D11Impl.cpp:30:
In file included from D:/Repos/funeral/extlibs/DiligentCore/Graphics/GraphicsEngineD3D11/include\ShaderResourceBindingD3D11Impl.hpp:37:
D:/Repos/funeral/extlibs/DiligentCore/Graphics/GraphicsEngineD3D11/include\ShaderResourceCacheD3D11.hpp:567:41: error: explicit instantiation of undefined function template 'TransitionResourceStates'
  567 | template void ShaderResourceCacheD3D11::TransitionResourceStates<ShaderResourceCacheD3D11::StateTransitionMode::Transition>(DeviceContextD3D11Impl& Ctx);
      |                                         ^
D:/Repos/funeral/extlibs/DiligentCore/Graphics/GraphicsEngineD3D11/include\ShaderResourceCacheD3D11.hpp:378:10: note: explicit instantiation refers here
  378 |     void TransitionResourceStates(DeviceContextD3D11Impl& Ctx);
      |          ^
D:/Repos/funeral/extlibs/DiligentCore/Graphics/GraphicsEngineD3D11/include\ShaderResourceCacheD3D11.hpp:568:41: error: explicit instantiation of undefined function template 'TransitionResourceStates'
  568 | template void ShaderResourceCacheD3D11::TransitionResourceStates<ShaderResourceCacheD3D11::StateTransitionMode::Verify>(DeviceContextD3D11Impl& Ctx);
      |                                         ^
D:/Repos/funeral/extlibs/DiligentCore/Graphics/GraphicsEngineD3D11/include\ShaderResourceCacheD3D11.hpp:378:10: note: explicit instantiation refers here
  378 |     void TransitionResourceStates(DeviceContextD3D11Impl& Ctx);
      |          ^
2 errors generated.```

File: Graphics/GraphicsEngineD3D11/include/ShaderVariableManagerD3D11.hpp (hunk 1 above)
Comment (Sixmorphugus) [reply Graphics/GraphicsEngineD3D11/include/ShaderVariableManagerD3D11.hpp:141]: It is, but clang emits a warning if `inline` is used here regardless, which is treated as a compile error by `-Werror`:

```
In file included from D:/Repos/[game]/extlibs/DiligentCore/Graphics/GraphicsEngineD3D11/src/ShaderResourceBindingD3D11Impl.cpp:30:
In file included from D:/Repos/[game]/extlibs/DiligentCore/Graphics/GraphicsEngineD3D11/include\ShaderResourceBindingD3D11Impl.hpp:38:
D:/Repos/[game]/extlibs/DiligentCore/Graphics/GraphicsEngineD3D11/include\ShaderVariableManagerD3D11.hpp:141:21: error: inline function 'Diligent::ShaderVariableManagerD3D11::ConstBuffBindInfo::BindResource' is not defined [-Werror,-Wundefined-inline]
  141 |         inline void BindResource(const BindResourceInfo& BindInfo);
      |                     ^
D:/Repos/[game]/extlibs/DiligentCore/Graphics/GraphicsEngine/include\ShaderResourceVariableBase.hpp:671:43: note: used here
  671 |         static_cast<ThisImplType*>(this)->BindResource(BindResourceInfo{pObject, Flags});
      |                                           ^
```

File: Graphics/GraphicsEngineD3DBase/include/EngineFactoryD3DBase.hpp (hunk 4 above)
Comment (Sixmorphugus) [reply Graphics/GraphicsEngineD3DBase/include/EngineFactoryD3DBase.hpp:113]: I see now, sorry about that will put it back.

**Comment by Sixmorphugus** [comment]
Squashed to clean up references to files that shouldn't have been edited.

File: Graphics/GraphicsEngineD3D11/include/ShaderResourceCacheD3D11.hpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (Sixmorphugus) [reply Graphics/GraphicsEngineD3D11/include/ShaderResourceCacheD3D11.hpp:386]: Trying something...

[the remark this answers is not in the document]
Comment (Sixmorphugus) [reply Graphics/GraphicsEngineD3D11/include/ShaderResourceCacheD3D11.hpp]: Ok, that worked! The instantiation had to be in CPP instead of HPP.

File: Graphics/GraphicsEngineD3D12/include/DescriptorHeap.hpp [hunk 8]
```diff
@@ -341,23 +341,23 @@ class DescriptorHeapAllocationManager
...
-
     // First CPU descriptor handle in the available descriptor range
     D3D12_CPU_DESCRIPTOR_HANDLE m_FirstCPUHandle = {0};
 
     // First GPU descriptor handle in the available descriptor range
     D3D12_GPU_DESCRIPTOR_HANDLE m_FirstGPUHandle = {0};
 
     size_t m_MaxAllocatedSize = 0;
+        
+    // Allocations manager used to handle descriptor allocations within the heap
```
Comment (TheMostDiligent) [inline Graphics/GraphicsEngineD3D12/include/DescriptorHeap.hpp:352]: Please don't change the order of members - the order in which they are destructed may be important

File: Graphics/GraphicsEngineD3D11/src/ShaderResourceCacheD3D11.cpp [hunk 9]
```diff
@@ -217,6 +207,21 @@ void ShaderResourceCacheD3D11::TransitionResources(DeviceContextD3D11Impl& Ctx,
...
+    
+    TransitionResources<Mode>(Ctx, static_cast<ID3D11Buffer*>(nullptr));
+    TransitionResources<Mode>(Ctx, static_cast<ID3D11ShaderResourceView*>(nullptr));
+    TransitionResources<Mode>(Ctx, static_cast<ID3D11SamplerState*>(nullptr));
+    TransitionResources<Mode>(Ctx, static_cast<ID3D11UnorderedAccessView*>(nullptr));
+}
+
+// Instantiate templates
+template void ShaderResourceCacheD3D11::TransitionResourceStates<ShaderResourceCacheD3D11::StateTransitionMode::Transition>(DeviceContextD3D11Impl& Ctx);
+template void ShaderResourceCacheD3D11::TransitionResourceStates<ShaderResourceCacheD3D11::StateTransitionMode::Verify>(DeviceContextD3D11Impl& Ctx);
```
Comment (TheMostDiligent) [inline Graphics/GraphicsEngineD3D11/src/ShaderResourceCacheD3D11.cpp:224]: Please don't move the function in the file - it only creates extra delta

File: Graphics/GraphicsEngineD3D11/include/ShaderVariableManagerD3D11.hpp (hunk 1 above)
Comment (TheMostDiligent) [reply Graphics/GraphicsEngineD3D11/include/ShaderVariableManagerD3D11.hpp:141]: OK, looks like `__forceinline` indeed needs to be removed here.

**Review by TheMostDiligent — CHANGES_REQUESTED** [review]

File: Graphics/GraphicsEngineD3D11/src/ShaderResourceCacheD3D11.cpp (hunk 9 above)
Comment (Sixmorphugus) [reply Graphics/GraphicsEngineD3D11/src/ShaderResourceCacheD3D11.cpp:224]: It doesn't compile if put into the .hpp due to the definition only being in .cpp.
No idea why it works in MSVC.

File: Graphics/GraphicsEngineD3D12/include/DescriptorHeap.hpp (hunk 8 above)
Comment (Sixmorphugus) [reply Graphics/GraphicsEngineD3D12/include/DescriptorHeap.hpp:352]: OK, will revert.

File: Graphics/GraphicsEngineD3D11/src/ShaderResourceCacheD3D11.cpp (hunk 9 above)
Comment (TheMostDiligent) [reply Graphics/GraphicsEngineD3D11/src/ShaderResourceCacheD3D11.cpp:224]: I mean, you moved the `ShaderResourceCacheD3D11::TransitionResourceStates` function as is in the `ShaderResourceCacheD3D11.cpp` file, which created extra changes.

Comment (Sixmorphugus) [reply Graphics/GraphicsEngineD3D11/src/ShaderResourceCacheD3D11.cpp:224]: Oh, let me move it back to where it was.

Comment (Sixmorphugus) [reply Graphics/GraphicsEngineD3D11/src/ShaderResourceCacheD3D11.cpp]: Moved back.

**Comment by TheMostDiligent** [comment]
The changes [break something](https://github.com/DiligentGraphics/DiligentCore/actions/runs/9669577868/job/26676510557?pr=495#step:8:48)

**Comment by Sixmorphugus** [comment]
Alright, will get it fixed tomorrow.

File: Graphics/GraphicsEngineD3D12/src/TextureD3D12Impl.cpp [hunk 10]
```diff
@@ -53,6 +53,7 @@ static DXGI_FORMAT GetClearFormat(DXGI_FORMAT Fmt, D3D12_RESOURCE_FLAGS Flags)
             case DXGI_FORMAT_R16_TYPELESS:      return DXGI_FORMAT_D16_UNORM;
             case DXGI_FORMAT_R24G8_TYPELESS:    return DXGI_FORMAT_D24_UNORM_S8_UINT;
             case DXGI_FORMAT_R32G8X24_TYPELESS: return DXGI_FORMAT_D32_FLOAT_S8X24_UINT;
+            default: break;
```
Comment (TheMostDiligent) [inline Graphics/GraphicsEngineD3D12/src/TextureD3D12Impl.cpp:56]: Let's do
```cpp
default: return Fmt;
```
to make the logic more clear here

File: Graphics/GraphicsEngineD3D12/src/TextureD3D12Impl.cpp [hunk 11]
```diff
@@ -74,6 +75,7 @@ static DXGI_FORMAT GetClearFormat(DXGI_FORMAT Fmt, D3D12_RESOURCE_FLAGS Flags)
             case DXGI_FORMAT_R8_TYPELESS:           return DXGI_FORMAT_R8_UNORM;
             case DXGI_FORMAT_B8G8R8A8_TYPELESS:     return DXGI_FORMAT_B8G8R8A8_UNORM;
             case DXGI_FORMAT_B8G8R8X8_TYPELESS:     return DXGI_FORMAT_B8G8R8X8_UNORM;
+            default: break;
```
Comment (TheMostDiligent) [inline Graphics/GraphicsEngineD3D12/src/TextureD3D12Impl.cpp:78]: `default: return Fmt;`

File: Graphics/GraphicsEngineD3DBase/include/EngineFactoryD3DBase.hpp [hunk 12]
```diff
@@ -300,7 +305,7 @@ class EngineFactoryD3DBase : public EngineFactoryBase<BaseInterface>
     }
 
 private:
-    template <RENDER_DEVICE_TYPE DevType>
+    template <RENDER_DEVICE_TYPE MyDevType>
```
Comment (TheMostDiligent) [inline Graphics/GraphicsEngineD3DBase/include/EngineFactoryD3DBase.hpp:308]: Is this really needed?

**Review by TheMostDiligent — CHANGES_REQUESTED** [review]

File: Graphics/GraphicsEngineD3DBase/include/EngineFactoryD3DBase.hpp (hunk 12 above)
Comment (Sixmorphugus) [reply Graphics/GraphicsEngineD3DBase/include/EngineFactoryD3DBase.hpp:308]: Clang generates a name shadowing warning otherwise.

**Comment by Sixmorphugus** [comment]
wait did i mess something up? i was trying to rebase my stuff onto the new master commit

**Comment by TheMostDiligent** [comment]
I think if you push new commit, it can be reopened.
If not, just open another PR.

**Comment by Sixmorphugus** [comment]
Sorry about that, new PR [here](https://github.com/DiligentGraphics/DiligentCore/pull/498)
